File: backend/app/indicators/timeframes.py
"""Per-indicator timeframe classification.

The 12 indicators span three horizons. Surfacing this as a chip on each
pros/cons row lets the operator answer "which horizon is this signal
talking about?" at a glance without reading the indicator docs.

Single source of truth: backend authoritative, frontend mirrors via the
API response shape.

Mapping rationale (matches CLAUDE.md "Core Indicators (12)"):

* short  — same-day or week-to-week tactical signals: RSI(14),
  Bollinger band reversion, MACD cross, today's volume anomaly,
  today's A/D Day Count, current VIX level.
* mid    — multi-week trend / regime: 50-day MA distance, relative
  strength vs SPX over a few weeks, the 50/200 SMA cross of SPX itself.
* long   — multi-month / multi-quarter macro: 200-day MA dominance,
  10Y-2Y yield spread (recession indicator), DXY (USD strength regime),
  Fed Funds Rate (policy regime).

Each indicator's primary horizon is captured. Indicators that touch
more than one timeframe (e.g. RSI uses both daily and weekly windows)
get the *dominant* horizon — the one that drives the buy/sell call.
"""

from __future__ import annotations

from typing import Final, Literal

Timeframe = Literal["short", "mid", "long"]

INDICATOR_TIMEFRAMES: Final[dict[str, Timeframe]] = {
    # Per-ticker direction (4)
    "price_vs_ma": "mid",
    "rsi": "short",
    "volume_anomaly": "short",
    "relative_strength": "mid",
    # Market regime (4)
    "spx_ma": "mid",
    "ad_day": "short",
    "vix": "short",
    "yield_spread": "long",
    # Timing (2)
    "macd": "short",
    "bollinger": "short",
    # Macro (2)
    "dxy": "long",
    "fed_rate": "long",
}


def timeframe_for(indicator_name: str) -> Timeframe:
    """Look up the timeframe for an indicator NAME.

    Raises ``KeyError`` for unknown names — callers (which hold the
    indicator NAME constants directly) should not be guessing.
    """
    return INDICATOR_TIMEFRAMES[indicator_name]


__all__ = ("INDICATOR_TIMEFRAMES", "Timeframe", "timeframe_for")

File: frontend/src/lib/timeframes.ts
// Display layer for the indicator timeframe chip. Backend
// (`app/indicators/timeframes.py`) is the single source of truth for
// the mapping; this module just renders short/mid/long.

export type Timeframe = 'short' | 'mid' | 'long';

interface TimeframeStyle {
  label: string;
  className: string;
  ariaLabel: string;
}

export const TIMEFRAME_STYLE: Record<Timeframe, TimeframeStyle> = {
  short: {
    label: '短期',
    className: 'border-sky-500/40 bg-sky-500/10 text-sky-300',
    ariaLabel: '短期訊號',
  },
  mid: {
    label: '中期',
    className: 'border-violet-500/40 bg-violet-500/10 text-violet-300',
    ariaLabel: '中期訊號',
  },
  long: {
    label: '長期',
    className: 'border-teal-500/40 bg-teal-500/10 text-teal-300',
    ariaLabel: '長期訊號',
  },
};

// Mirror of backend ``app/indicators/timeframes.py::INDICATOR_TIMEFRAMES``.
// Authoritative source is the backend (and the API ships ``timeframe`` on
// every ProsConsItem); this map exists for components that render an
// indicator row directly without a ProsConsItem in hand — e.g. the
// per-ticker IndicatorRow and the MarketRegimeIndicatorList.
export const INDICATOR_TIMEFRAMES: Record<string, Timeframe> = {
  price_vs_ma: 'mid',
  rsi: 'short',
  volume_anomaly: 'short',
  relative_strength: 'mid',
  spx_ma: 'mid',
  ad_day: 'short',
  vix: 'short',
  yield_spread: 'long',
  macd: 'short',
  bollinger: 'short',
  dxy: 'long',
  fed_rate: 'long',
};
